http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/AwsCrtHttpClientBase.java

```diff
@@ -0,0 +1,180 @@
+ */
+
+package software.amazon.awssdk.http.crt.internal;
+
+import static software.amazon.awssdk.crtcore.CrtConfigurationUtils.resolveHttpMonitoringOptions;
+import static software.amazon.awssdk.crtcore.CrtConfigurationUtils.resolveProxy;
+import static software.amazon.awssdk.http.SdkHttpConfigurationOption.PROTOCOL;
+import static software.amazon.awssdk.http.crt.internal.AwsCrtConfigurationUtils.buildSocketOptions;
+import static software.amazon.awssdk.http.crt.internal.AwsCrtConfigurationUtils.resolveCipherPreference;
+import static software.amazon.awssdk.utils.FunctionalUtils.invokeSafely;
+
+import java.net.URI;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.CrtResource;
+import software.amazon.awssdk.crt.http.HttpClientConnectionManager;
+import software.amazon.awssdk.crt.http.HttpClientConnectionManagerOptions;
+import software.amazon.awssdk.crt.http.HttpMonitoringOptions;
+import software.amazon.awssdk.crt.http.HttpProxyOptions;
+import software.amazon.awssdk.crt.io.ClientBootstrap;
+import software.amazon.awssdk.crt.io.SocketOptions;
+import software.amazon.awssdk.crt.io.TlsContext;
+import software.amazon.awssdk.crt.io.TlsContextOptions;
+import software.amazon.awssdk.http.Protocol;
+import software.amazon.awssdk.http.SdkHttpConfigurationOption;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+/**
+ * Common functionality and configuration for the CRT Http clients.
+ */
+@SdkInternalApi
+public abstract class AwsCrtHttpClientBase implements SdkAutoCloseable {
+    private static final Logger log = Logger.loggerFor(AwsCrtHttpClientBase.class);
+
+    private static final String AWS_COMMON_RUNTIME = "AwsCommonRuntime";
+    private static final long DEFAULT_STREAM_WINDOW_SIZE = 16L * 1024L * 1024L; // 16 MB
+
+    protected final long readBufferSize;
+    private final Map<URI, HttpClientConnectionManager> connectionPools = new ConcurrentHashMap<>();
+    private final LinkedList<CrtResource> ownedSubResources = new LinkedList<>();
+    private final ClientBootstrap bootstrap;
+    private final SocketOptions socketOptions;
+    private final TlsContext tlsContext;
+    private final HttpProxyOptions proxyOptions;
+    private final HttpMonitoringOptions monitoringOptions;
+    private final long maxConnectionIdleInMilliseconds;
+    private final int maxConnectionsPerEndpoint;
+    private boolean isClosed = false;
+
+    protected AwsCrtHttpClientBase(AwsCrtClientBuilderBase builder, AttributeMap config) {
+        if (config.get(PROTOCOL) == Protocol.HTTP2) {
+            throw new UnsupportedOperationException("HTTP/2 is not supported in AwsCrtAsyncHttpClient yet. Use "
```

Comment:
nit: this mentions the async CRT client, but we're throwing from the abstract base. Does this line need to be more generic or moved to the async client?

Reply:
I have made it more generic.